lamindb_setup/core/upath.py

```diff
@@ -421,6 +424,48 @@ def view_tree(
     logger.print(message)
 
 
+# TODO: replace with something faster at some point
+# also consolidate with what's in _settings_storage
+def get_bucket_region(bucket_name) -> str:
+    """Retrieves the region for a given S3 bucket using boto3.
+
+    Args:
+    - bucket_name: Name of the S3 bucket.
+
+    Returns:
+    - The region in which the S3 bucket is located.
+    """
+    import boto3
+
+    s3 = boto3.client("s3")
+    bucket_location = s3.get_bucket_location(Bucket=bucket_name)
+    region = bucket_location["LocationConstraint"]
+    return region
+
+
+def to_url(upath):
+    """Public storage URL.
+
+    Generates a public URL for an object in an S3 bucket using fsspec's UPath,
+    considering the bucket's region.
+
+    Args:
+    - upath: A UPath object representing an S3 path.
+
+    Returns:
+    - A string containing the public URL to the S3 object.
+    """
+    if upath.protocol != "s3":
+        raise ValueError("The provided UPath must be an S3 path.")
+    bucket = upath._url.netloc
+    key = "/".join(upath.parts[1:])
+    region = get_bucket_region(bucket)
```

Comment:
s3fs fs should have this information, i would not use it like this, we should avoid creating new client randomly, we need to use the underlying s3fs fs clinet.

Reply:
Fixed in the commit above!

Reply:
There is also fs.url for that https://s3fs.readthedocs.io/en/latest/api.html?highlight=url#s3fs.core.S3FileSystem.url

Reply:
Hmm, this is what I took from your code in StorageSettings. 🤔

Reply:
- https://github.com/laminlabs/lamindb-setup/issues/696